codegen/core/src/es/event/mod.rs

```diff
@@ -0,0 +1,518 @@
+    ///
+    /// Checks uniqueness of all [`Event::name`][0]s and [`Event::version`][1]s.
+    ///
+    /// # Panics
+    ///
+    /// If some enum [`Variant`]s don't have exactly 1 [`Field`] and not marked
+    /// with `#[event(skip)]`. Checked by [`TryFrom`] impl  for [`Definition`].
+    ///
+    /// [0]: arcana_core::es::event::Event::name()
+    /// [1]: arcana_core::es::event::Event::version()
+    /// [`Field`]: syn::Field
+    /// [`Variant`]: syn::Variant
+    #[must_use]
+    pub fn unique_event_name_and_ver(&self) -> TokenStream {
+        let name = &self.ident;
+        let (impl_generics, ty_generics, where_clause) =
+            self.generics.split_for_impl();
+        let (event_sizes, event_array_population): (
+            Vec<TokenStream>,
+            TokenStream,
+        ) = self
+            .variants
+            .iter()
+            .filter_map(|(variant, attr)| {
+                attr.skip.is_none().then(|| {
+                    let ty = &variant.fields.iter().next().unwrap().ty;
+                    (
+                        quote! {
+                            <#ty as ::arcana::codegen::UniqueEvents>::COUNT
+                        },
+                        quote! {{
+                            let ev = #ty::__arcana_events();
+                            let mut local = 0;
+                            while local < ev.len() {
+                                res[global] = ev[local];
+                                local += 1;
+                                global += 1;
+                            }
+                        }},
+                    )
+                })
+            })
+            .unzip();
+
+        let event_sizes = event_sizes
+            .into_iter()
+            .fold(None, |acc, size| {
+                Some(acc.map(|acc| quote! { #acc + #size }).unwrap_or(size))
+            })
+            .unwrap_or(quote! { 1 });
+
+        quote! {
+            #[automatically_derived]
+            impl #impl_generics ::arcana::codegen::UniqueEvents for
+                #name #ty_generics #where_clause
+            {
+                const COUNT: usize = #event_sizes;
+            }
+
+            impl #impl_generics #name #ty_generics #where_clause {
+                #[automatically_derived]
+                #[doc(hidden)]
+                pub const fn __arcana_events() -> [
+                    (&'static str, u16);
+                    <Self as ::arcana::codegen::UniqueEvents>::COUNT
+                ] {
+                    let mut res = [
+                        ("", 0);
+                        <Self as ::arcana::codegen::UniqueEvents>::COUNT
+                    ];
+
+                    let mut global = 0;
+
+                    #event_array_population
+
+                    res
+                }
+            }
+
+            ::arcana::codegen::sa::const_assert!(
+                !::arcana::codegen::unique_events::has_duplicates(
+                    #name::__arcana_events()
```

Comment:
That simply won't work with generics.

Reply:
@tyranron it looks like there is no way to make it work with generics for now 😔
With your solution compiler error is

> can't use generic parameters from outer function [E0401](https://doc.rust-lang.org/error-index.html#E0401)
> use of generic parameter from outer function